probe-rs/src/probe/daplink/tools.rs

```diff
@@ -140,12 +146,18 @@ pub fn open_v2_device(device: Device<rusb::Context>) -> Option<DAPLinkDevice> {
 
 /// Attempt to open the given DebugProbeInfo in CMSIS-DAP v2 mode if possible,
 /// otherwise in v1 mode.
-pub fn open_device_from_info(info: &DebugProbeInfo) -> Option<DAPLinkDevice> {
+pub fn open_device_from_selector(
+    selector: impl Into<DebugProbeSelector>,
+) -> Result<DAPLinkDevice, ProbeCreationError> {
+    let selector = selector.into();
     // Try using rusb to open a v2 device. This might fail if
     // the device does not support v2 operation or due to driver
     // or permission issues with opening bulk devices.
     if let Ok(devices) = rusb::Context::new().and_then(|ctx| ctx.devices()) {
-        for device in devices.iter() {
+        for device in devices
+            .iter()
+            .filter(|device| get_daplink_info(device).is_some())
```

Comment:
Is there any reason to add this filter step? The loop body already only accepts an exact matching device, and calling `get_daplink_info` for each device only takes more time (since it has to do more work on each device to check the product string) but I don't think it gives any benefit.

Reply:
The old code would open ANY device with matching vid/pid/serial even if it was no daplink. this worked because we would always pass the device type with the device info we used to identify the probe. but now i added a mechanism where the user can open a device with only the vid/pid, without specifying a type. so it will always open a daplink, even for a connected jlink/stlink because the vid/pid would match. so i added this check for the devicetype :)

Reply:
Ah, I see! I think there might be a simpler way to structure this here, I'll have a look.